capsules/src/kv_store.rs

```diff
@@ -0,0 +1,426 @@
+//! Tock TicKV capsule.
+//!
+//! This capsule implements the TicKV library in Tock. This is done
+//! using the TicKV library (libraries/tickv).
+//!
+//! This capsule interfaces with flash and exposes the Tock `hil::kv_system`
+//! interface to others.
+//!
+//! +-----------------------+
+//! |                       |
+//! |  Capsule using K-V    |
+//! |                       |
+//! +-----------------------+
+//!
+//!    capsules::kv_store
+//!
+//! +-----------------------+
+//! |                       |
+//! | K-V store (this file)|
+//! |                       |
+//! +-----------------------+
+//!
+//!    hil::kv_system
+//!
+//! +-----------------------+
+//! |                       |
+//! |  K-V library          |
+//! |                       |
+//! +-----------------------+
+//!
+//!    hil::flash
+
+use core::cell::Cell;
+use kernel::collections::list::{List, ListLink, ListNode};
+use kernel::hil::kv_system::{self, KVSystem};
+use kernel::utilities::cells::{OptionalCell, TakeCell};
+use kernel::ErrorCode;
+
+#[derive(Clone, Copy, PartialEq)]
+enum Operation {
+    None,
+    Get,
+    Set,
+    Delete,
+}
+
+pub struct KVStore<'a, K: KVSystem<'a> + KVSystem<'a, K = T>, T: 'static + kv_system::KeyType> {
+    mux_kv: &'a MuxKVStore<'a, K, T>,
+    next: ListLink<'a, KVStore<'a, K, T>>,
+
+    client: OptionalCell<&'a dyn kv_system::StoreClient<T>>,
+
+    next_operation: Cell<Operation>,
+
+    hashed_key: TakeCell<'static, T>,
+    unhashed_key: TakeCell<'static, [u8]>,
+    value: TakeCell<'static, [u8]>,
+}
+
+impl<'a, K: KVSystem<'a, K = T>, T: kv_system::KeyType> ListNode<'a, KVStore<'a, K, T>>
+    for KVStore<'a, K, T>
+{
+    fn next(&self) -> &'a ListLink<KVStore<'a, K, T>> {
+        &self.next
+    }
+}
+
+impl<'a, K: KVSystem<'a, K = T>, T: kv_system::KeyType> KVStore<'a, K, T> {
+    pub fn new(mux_kv: &'a MuxKVStore<'a, K, T>, key: &'static mut T) -> KVStore<'a, K, T> {
+        Self {
+            mux_kv,
+            next: ListLink::empty(),
+            client: OptionalCell::empty(),
+            next_operation: Cell::new(Operation::None),
+            hashed_key: TakeCell::new(key),
+            unhashed_key: TakeCell::empty(),
+            value: TakeCell::empty(),
+        }
+    }
+
+    pub fn set_client(&self, client: &'a dyn kv_system::StoreClient<T>) {
+        self.client.set(client);
+    }
+
+    pub fn get(
+        &self,
+        unhashed_key: &'static mut [u8],
+        value: &'static mut [u8],
+    ) -> Result<(), (&'static mut [u8], &'static mut [u8], Result<(), ErrorCode>)> {
+        if self.mux_kv.operation.get() == Operation::None {
+            self.mux_kv.operation.set(Operation::Get);
+            if let Err((unhashed_key, hashed_key, e)) = self
+                .mux_kv
+                .kv
+                .generate_key(unhashed_key, self.hashed_key.take().unwrap())
```

Comment:
There are many `unwrap()`s in this file. Aside from this not being ideal from a "setting an example" perspective, doesn't this add many panics and increase the code size?